saga-core/src/main/java/org/apache/servicecomb/saga/core/ForwardRecovery.java

```diff
@@ -49,6 +49,15 @@ public SagaResponse apply(SagaTask task, SagaRequest request, SagaResponse paren
           ignored);
       throw new TransactionFailedException(ignored);
     }
+    throw new TransactionAbortedException(
+        String.format(
+            "Too many failures in transaction %s of service %s, abort the transaction!",
+            request.transaction(),
+            request.serviceName()));
+  }
+
+  private boolean isRetryable(int i, Transaction transaction) {
+    return transaction.retries() == Transaction.INFINITE_RETRY || i <= transaction.retries();
```

Comment:
I think it could be an issue here. if the transaction.retries() is 0, it just run sout of the loop and has no retry at all. But you mention in the document

> If this parameter is less than or equal to 0, transaction will retry infinitely.

Reply:
Thanks @KomachiSion , as i starts from 0
> for(int i = 0; isRetryable(i, request.transaction()); i++)
it looks like the condition should be i < transcaction.retries() ?